File: Stacks/LinkedlistStack.cpp
#include <bits/stdc++.h>
using namespace std;

template <class T>
class LinkedListStack
{

    class Node
    {
    private:
        T value;
        Node *next;
        Node *prev;

    public:
        Node(T value, Node *prev = nullptr, Node *next = nullptr)
        {
            this->value = value;
            this->prev = prev;
            this->next = next;
        }

        void setPrev(Node *prevprev) { this->prev = prevprev; }
        void setNext(Node *next) { this->next = next; }
        void setValue(T value) { this->value = value; }
        Node *getNext() { return this->next; }
        Node *getPrev() { return this->prev; }
        T getValue() { return this->value; }
    };

    Node *tail;
    int size;

public:
    LinkedListStack()
    {
        this->tail = nullptr;
        this->size = 0;
    }

    Node *getTail() { return this->tail; }
    int getSize() { return this->size; }

    void setTail(Node *newTail) { this->tail = newTail; }
    void setSize(int newSize) { this->size = newSize; }
    void increaseSize() { this->size++; }
    void decreaseSize() { this->size--; }

    bool isEmpty() { return this->size == 0; }

    void push(T value)
    {
        Node *newNode = new Node(value, this->tail, nullptr);

        if (!isEmpty()) this->tail->setNext(newNode);
        this->tail = newNode;
        increaseSize();
    }

    T pop()
    {
        if (isEmpty()) throw "List is empty";

        T value = this->tail->getValue();
        Node* q = this->tail;

        if(size > 1) this->tail->getPrev()->setNext(nullptr);
        this->tail = this->tail->getPrev();

        decreaseSize();

        delete q;
        return value;
    }

    T top()
    {
        if (isEmpty()) throw "List is empty";
        return this->tail->getValue();
    }
};

File: Stacks/LinkedListStack.java
public class LinkedListStack<E> implements Stack<E> {

    SinglyLinkedList<E> list = new SinglyLinkedList<>();

    /* ------------------------ Stack Methods ------------------------ */

    @Override
    public void push(E e) {
        list.addFirst(e);
    }

    @Override
    public E pop() {
        return list.removeFirst();
    }

    @Override
    public E top() {
        return list.first();
    }

    @Override
    public int size() {
        return list.size();
    }

    @Override
    public boolean isEmpty() {
        return list.isEmpty();
    }

    public static boolean isMatched(String expression) {
        final String opening = "({[";
        final String closing = ")}]";
        Stack<Character> buffer = new LinkedListStack<>();
        for (char c : expression.toCharArray()) {
            if (opening.indexOf(c) != -1) {
                buffer.push(c);
            } else if (closing.indexOf(c) != -1) {
                if (buffer.isEmpty()) return false;
                if (closing.indexOf(c) != opening.indexOf(buffer.pop())) return false;
            }
        }
        return buffer.isEmpty();
    }

    public static boolean isHTMLMatched(String html) {
        Stack<String> buffer = new LinkedListStack<>();
        int j = html.indexOf('<');
        while (j != -1) {
            int k = html.indexOf('>', j + 1);
            if (k == -1) return false;
            String tag = html.substring(j + 1, k);
            if (!tag.startsWith("/")) {
                buffer.push(tag);
            } else {
                if (buffer.isEmpty()) return false;
                if (!tag.substring(1).equals(buffer.pop())) return false;
            }
            j = html.indexOf('<', k + 1);
        }
        return buffer.isEmpty();
    }

    /* ------------------------ LinkedList Class ------------------------ */

    public static class SinglyLinkedList<T> {

        private Node head;
        private Node tail;
        private int size;

        public SinglyLinkedList() {
            this.head = null;
            this.tail = null;
            this.size = 0;
        }

        /* ------------------------ LinkedList Methods ------------------------ */

        public int size() {
            return size;
        }

        public boolean isEmpty() {
            return size == 0;
        }

        public T first() {
            if (isEmpty()) throw new IllegalStateException("LinkedList is empty");
            return this.head.getValue();
        }

        public T last() {
            if (isEmpty()) throw new IllegalStateException("LinkedList is empty");
            return this.tail.getValue();
        }

        public void addFirst(T value) {
            head = new Node(value, head);
            if (isEmpty()) tail = head;
            size++;
        }

        public void addLast(T value) {
            Node newNode = new Node(value, null);
            if (isEmpty()) head = newNode;
            else tail.setNext(newNode);
            tail = newNode;
            size++;
        }

        public T removeFirst() {
            if (isEmpty()) throw new IllegalStateException("LinkedList is empty");
            T first = head.getValue();
            head = head.getNext();
            size--;
            return first;
        }

        public void traverse() {
            Node node = head;
            while (node != null) {
                System.out.println(node.getValue());
                node = node.getNext();
            }
        }

        /* ------------------------ Getter/Setter ------------------------ */

        public Node getHead() {
            return head;
        }

        public void setHead(Node head) {
            this.head = head;
        }

        public Node getTail() {
            return tail;
        }

        public void setTail(Node tail) {
            this.tail = tail;
        }

        public int getSize() {
            return size;
        }

        public void setSize(int size) {
            this.size = size;
        }

        /* ------------------------ Node Class ------------------------ */

        private class Node {
            private final T value;
            private Node next;

            public Node(T v, Node n) {
                value = v;
                next = n;
            }

            public T getValue() {
                return value;
            }

            public Node getNext() {
                return next;
            }

            public void setNext(Node next) {
                this.next = next;
            }
        }
    }
}
